// @roots/bud-client/lib/components/indicator/indicator.component.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Component = void 0;
const indicator_pulse_cjs_1 = require("./indicator.pulse.cjs");
/**
 * Indicator web component
 * @public
 */
class Component extends HTMLElement {
    constructor() {
        super(...arguments);
        /**
         * Component name
         * @public
         */
        this.name = `bud-activity-indicator`;
        /**
         * Status indicator colors
         * @public
         */
        this.colors = {
            success: [4, 120, 87, 1],
            error: [220, 38, 38, 1],
            warn: [252, 211, 77, 1],
            pending: [59, 130, 246, 1],
        };
    }
    /**
     * Root div querySelector selector
     * @public
     */
    get selector() {
        return `.${this.name}`;
    }
    /**
     * Get accessor: has errors
     * @public
     */
    get hasErrors() {
        return this.getAttribute('has-errors') == 'true';
    }
    /**
     * Get accessor: has warnings
     * @public
     */
    get hasWarnings() {
        return this.getAttribute('has-warnings') == 'true';
    }
    /**
     * Render status indicator
     * @public
     */
    renderShadow() {
        const container = document.createElement('div');
        container.classList.add(this.name);
        container.innerHTML = `
    <style>
    .bud-activity-indicator {
      position: fixed;
      width: 10px;
      height: 10px;
      left: 10px;
      bottom: 10px;
      z-index: 9999;
      margin: 5px;
      padding: 5px;
      -webkit-transition:
        all .6s ease-in-out,
      transition:
        all .6s ease-in-out;
      animation-fill-mode: forwards;
      pointer-events: none;
      border-radius: 50%;
      transform: scale(0);
      opacity: 0;
    }

    .show {
      opacity: 1;
      background-color: rgba(255, 255, 255, 1);
      transform: scale(1);
      transition:
        all .6s ease-in-out;
    }

    ${(0, indicator_pulse_cjs_1.pulse)(`success`, this.colors.success)}
    ${(0, indicator_pulse_cjs_1.pulse)(`error`, this.colors.error)}
    ${(0, indicator_pulse_cjs_1.pulse)(`warning`, this.colors.warn)}
    ${(0, indicator_pulse_cjs_1.pulse)(`pending`, this.colors.pending)}

    </style>
    `;
        this.attachShadow({ mode: 'open' }).appendChild(container);
    }
    /**
     * Show status indicator
     * @public
     */
    show() {
        this.hideTimeout && clearTimeout(this.hideTimeout);
        this.shadowRoot.querySelector(this.selector).classList.add('show');
    }
    /**
     * Hide status indicator
     */
    hide() {
        this.hideTimeout = setTimeout(() => {
            this.shadowRoot.querySelector(this.selector).classList.remove('show');
        }, 2000);
    }
    /**
     * Status is pending
     * @public
     */
    onPending() {
        this.show();
        this.shadowRoot
            .querySelector(this.selector)
            .classList.remove(`error`, `warning`, `success`);
        this.shadowRoot.querySelector(this.selector).classList.add('pending');
        this.hide();
    }
    /**
     * Status is success
     * @public
     */
    onSuccess() {
        this.show();
        this.shadowRoot
            .querySelector(this.selector)
            .classList.remove(`error`, `warning`, `pending`);
        this.shadowRoot.querySelector(this.selector).classList.add(`success`);
        this.hide();
    }
    /**
     * Status is error
     * @public
     */
    onError() {
        this.show();
        this.shadowRoot
            .querySelector(this.selector)
            .classList.remove(`warning`, `success`, `pending`);
        this.shadowRoot.querySelector(this.selector).classList.add(`error`);
    }
    /**
     * Status is warning
     * @public
     */
    onWarning() {
        this.show();
        this.shadowRoot
            .querySelector(this.selector)
            .classList.remove(`error`, `success`, `pending`);
        this.shadowRoot.querySelector(this.selector).classList.add(`warning`);
    }
    static get observedAttributes() {
        return ['has-errors', 'has-warnings', 'action'];
    }
    attributeChangedCallback() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        if ((_b = (_a = this.payload) === null || _a === void 0 ? void 0 : _a.errors) === null || _b === void 0 ? void 0 : _b.length)
            return this.onError();
        if ((_d = (_c = this.payload) === null || _c === void 0 ? void 0 : _c.warnings) === null || _d === void 0 ? void 0 : _d.length)
            return this.onWarning();
        if (!((_f = (_e = this.payload) === null || _e === void 0 ? void 0 : _e.errors) === null || _f === void 0 ? void 0 : _f.length) &&
            !((_h = (_g = this.payload) === null || _g === void 0 ? void 0 : _g.warnings) === null || _h === void 0 ? void 0 : _h.length) &&
            this.payload.action == 'built')
            return this.onSuccess();
        if (((_j = this.payload) === null || _j === void 0 ? void 0 : _j.action) == 'building' ||
            ((_k = this.payload) === null || _k === void 0 ? void 0 : _k.action) == 'sync')
            return this.onPending();
    }
    connectedCallback() {
        if (!this.rendered) {
            this.renderShadow();
            this.rendered = true;
        }
    }
}
exports.Component = Component;
