// @roots/bud-framework/lib/logger/logger.constants.js
import figures from 'figures';
/**
 * Instance configuration
 *
 * @internal
 */
export const INSTANCE_CONFIG = {
    displayScope: true,
    displayBadge: true,
    displayDate: false,
    displayFilename: false,
    displayLabel: false,
    displayTimestamp: false,
    underlineLabel: false,
    underlineMessage: false,
    underlinePrefix: false,
    underlineSuffix: false,
    uppercaseLabel: false,
};
export const types = app => ({
    error: {
        badge: figures.cross,
        color: 'red',
        label: 'error',
        logLevel: "error" /* LEVEL.ERROR */,
    },
    fatal: {
        badge: figures.cross,
        color: 'red',
        label: 'fatal',
        logLevel: "error" /* LEVEL.ERROR */,
    },
    warn: {
        badge: figures.warning,
        color: 'yellow',
        label: 'warning',
        logLevel: "error" /* LEVEL.ERROR */,
    },
    star: {
        badge: figures.star,
        color: 'cyan',
        label: 'instantiate',
        logLevel: "timer" /* LEVEL.STANDARD */,
    },
    info: {
        badge: figures.info,
        color: 'magenta',
        label: 'log',
        logLevel: "log" /* LEVEL.VERBOSE */,
    },
    success: {
        badge: figures.tick,
        color: 'green',
        label: 'success',
        logLevel: "timer" /* LEVEL.STANDARD */,
    },
    complete: {
        badge: figures.circleFilled,
        color: 'cyan',
        label: 'complete',
        logLevel: "timer" /* LEVEL.STANDARD */,
    },
    pending: {
        badge: figures.ellipsis,
        color: 'cyan',
        label: 'pending',
        logLevel: "timer" /* LEVEL.STANDARD */,
        stream: process.stdout,
    },
    note: {
        badge: figures.bullet,
        color: 'blue',
        label: 'note',
        logLevel: "timer" /* LEVEL.STANDARD */,
    },
    start: {
        badge: figures.play,
        color: 'green',
        label: 'start',
        logLevel: "timer" /* LEVEL.STANDARD */,
    },
    pause: {
        badge: figures.squareSmallFilled,
        color: 'yellow',
        label: 'pause',
        logLevel: "timer" /* LEVEL.STANDARD */,
    },
    debug: {
        badge: figures.circleFilled,
        color: 'red',
        label: 'log',
        logLevel: "log" /* LEVEL.VERBOSE */,
    },
    await: {
        badge: figures.ellipsis,
        color: 'cyan',
        label: 'awaiting',
        logLevel: "timer" /* LEVEL.STANDARD */,
    },
    watch: {
        badge: figures.ellipsis,
        color: 'yellow',
        label: 'watching',
        logLevel: "timer" /* LEVEL.STANDARD */,
    },
    log: {
        badge: figures.pointer,
        color: 'blue',
        label: 'log',
        logLevel: "timer" /* LEVEL.STANDARD */,
    },
});
