// @roots/bud-compiler/lib/Reporter/BudError.js
import { __decorate } from "tslib";
import { bind } from 'helpful-decorators';
export default class BudError {
    constructor(statsError) {
        this.statsError = statsError;
        this.line = 0;
        this.column = 0;
        this.type = null;
        if (!statsError)
            return;
        this.source =
            statsError?.moduleName?.split('!').pop() ??
                statsError?.moduleIdentifier?.split('!').pop() ??
                null;
        this.setFile(this.source);
        this.message = statsError?.message ?? null;
    }
    hasType() {
        return typeof this.type === 'string';
    }
    getType() {
        return this.type;
    }
    setType(type) {
        this.type = type;
    }
    getFile() {
        return this.file;
    }
    setFile(file) {
        this.file = typeof file === 'function' ? file(this.file ?? '') : file;
        return this;
    }
    hasFile() {
        return typeof this.file === 'string';
    }
    getMessage() {
        return this.message;
    }
    setMessage(message) {
        this.message =
            typeof message === 'function' ? message(this.message ?? '') : message;
        return this;
    }
    hasMessage() {
        return typeof this.message === 'string';
    }
    getLine() {
        return this.line;
    }
    setLine(line) {
        this.line = typeof line === 'string' ? Number.parseInt(line) : line;
        return this;
    }
    hasLine() {
        return this.line !== 0;
    }
    getColumn() {
        return this.column;
    }
    setColumn(column) {
        this.column =
            typeof column === 'string' ? Number.parseInt(column) : column;
        return this;
    }
    hasColumn() {
        return this.column !== 0;
    }
}
__decorate([
    bind
], BudError.prototype, "hasType", null);
__decorate([
    bind
], BudError.prototype, "getType", null);
__decorate([
    bind
], BudError.prototype, "setType", null);
__decorate([
    bind
], BudError.prototype, "getFile", null);
__decorate([
    bind
], BudError.prototype, "setFile", null);
__decorate([
    bind
], BudError.prototype, "hasFile", null);
__decorate([
    bind
], BudError.prototype, "getMessage", null);
__decorate([
    bind
], BudError.prototype, "setMessage", null);
__decorate([
    bind
], BudError.prototype, "hasMessage", null);
__decorate([
    bind
], BudError.prototype, "getLine", null);
__decorate([
    bind
], BudError.prototype, "setLine", null);
__decorate([
    bind
], BudError.prototype, "hasLine", null);
__decorate([
    bind
], BudError.prototype, "getColumn", null);
__decorate([
    bind
], BudError.prototype, "setColumn", null);
__decorate([
    bind
], BudError.prototype, "hasColumn", null);
